File: bindings/python/vietasr/pipeline.py
import json
from pathlib import Path
from typing import Optional, Union

from vietasr.ffi import native
from vietasr.result import Result
from vietasr.session import Session


class PipelineError(RuntimeError):
    pass


class Pipeline:
    def __init__(self, handle):
        self.handle = handle
        self.ffi = native.ffi

    @classmethod
    def preset(cls, name: str) -> "Pipeline":
        handle = native.lib.vietasr_pipeline_preset(name.encode("utf-8"))
        if handle == native.ffi.NULL:
            raise PipelineError(cls.last_error() or f"unknown preset: {name}")
        return cls(handle)

    @classmethod
    def new(cls) -> "Pipeline":
        return cls(native.lib.vietasr_pipeline_new())

    def add(self, module_name: str, config: Optional[dict] = None) -> "Pipeline":
        json_config = json.dumps(config or {})
        status = native.lib.vietasr_pipeline_add_module(
            self.handle,
            module_name.encode("utf-8"),
            json_config.encode("utf-8"),
        )
        if status != 0:
            raise PipelineError(self.last_error() or f"add({module_name}) failed: {status}")
        return self

    def set_backend(self, backend: str) -> "Pipeline":
        mapping = {"auto": 0, "onnx": 1, "coreml": 2}
        status = native.lib.vietasr_pipeline_set_backend(self.handle, mapping[backend])
        if status != 0:
            raise PipelineError(self.last_error() or f"set_backend({backend}) failed")
        return self

    def set_model_dir(self, model_dir: Union[str, Path]) -> "Pipeline":
        status = native.lib.vietasr_pipeline_set_model_dir(
            self.handle, str(model_dir).encode("utf-8")
        )
        if status != 0:
            raise PipelineError(self.last_error() or "set_model_dir failed")
        return self

    def build(self) -> "Pipeline":
        status = native.lib.vietasr_pipeline_build(self.handle)
        if status != 0:
            raise PipelineError(self.last_error() or "build failed")
        return self

    def transcribe(self, source: Union[str, Path, "np.ndarray"], sample_rate: float = 16000.0) -> Result:
        if isinstance(source, (str, Path)):
            raw = native.lib.vietasr_transcribe_file(self.handle, str(source).encode("utf-8"))
            if raw == native.ffi.NULL:
                raise PipelineError(self.last_error() or "transcribe_file failed")
            return Result.from_json(self.ffi.string(raw).decode("utf-8"))
        import numpy as np
        array = np.asarray(source).astype(np.int16)
        buffer = self.ffi.cast("const short*", self.ffi.from_buffer(array))
        raw = native.lib.vietasr_transcribe_buffer(self.handle, buffer, array.size, sample_rate)
        if raw == native.ffi.NULL:
            raise PipelineError(self.last_error() or "transcribe_buffer failed")
        return Result.from_json(self.ffi.string(raw).decode("utf-8"))

    def stream(self, sample_rate: float = 16000.0) -> Session:
        handle = native.lib.vietasr_session_new(self.handle, sample_rate)
        if handle == native.ffi.NULL:
            raise PipelineError(self.last_error() or "session creation failed")
        return Session(handle, self.ffi)

    @staticmethod
    def last_error() -> str:
        raw = native.lib.vietasr_last_error()
        if raw == native.ffi.NULL:
            return ""
        return native.ffi.string(raw).decode("utf-8")

    def close(self) -> None:
        if self.handle:
            native.lib.vietasr_pipeline_free(self.handle)
            self.handle = None

    def __del__(self):
        self.close()

File: bindings/csharp/Vietasr/Pipeline.cs
using System;
using System.Collections.Generic;
using System.Text.Json;

namespace Vietasr
{
    public sealed class Pipeline : IDisposable
    {
        private IntPtr handle;

        private Pipeline(IntPtr handle)
        {
            this.handle = handle;
        }

        public static Pipeline Preset(string name)
        {
            IntPtr handle = NativeMethods.vietasr_pipeline_preset(name);
            if (handle == IntPtr.Zero)
            {
                throw new VietasrException(LastError($"unknown preset: {name}"));
            }
            return new Pipeline(handle);
        }

        public static Pipeline Create()
        {
            return new Pipeline(NativeMethods.vietasr_pipeline_new());
        }

        public Pipeline Add(string moduleName, string? jsonConfig = null)
        {
            int status = NativeMethods.vietasr_pipeline_add_module(
                handle, moduleName, jsonConfig ?? "{}");
            if (status != 0)
            {
                throw new VietasrException(LastError($"add({moduleName}) failed: {status}"));
            }
            return this;
        }

        public Pipeline SetBackend(Backend backend)
        {
            int status = NativeMethods.vietasr_pipeline_set_backend(handle, (int)backend);
            if (status != 0)
            {
                throw new VietasrException(LastError("setBackend failed"));
            }
            return this;
        }

        public Pipeline SetModelDir(string dir)
        {
            int status = NativeMethods.vietasr_pipeline_set_model_dir(handle, dir);
            if (status != 0)
            {
                throw new VietasrException(LastError("setModelDir failed"));
            }
            return this;
        }

        public Pipeline Build()
        {
            int status = NativeMethods.vietasr_pipeline_build(handle);
            if (status != 0)
            {
                throw new VietasrException(LastError("build failed"));
            }
            return this;
        }

        public Result Transcribe(string wavPath)
        {
            IntPtr raw = NativeMethods.vietasr_transcribe_file(handle, wavPath);
            if (raw == IntPtr.Zero)
            {
                throw new VietasrException(LastError("transcribe failed"));
            }
            return Result.FromJson(NativeMethods.PtrToString(raw));
        }

        public Result Transcribe(short[] pcm, float sampleRate)
        {
            IntPtr raw = NativeMethods.vietasr_transcribe_buffer(
                handle, pcm, pcm.Length, sampleRate);
            if (raw == IntPtr.Zero)
            {
                throw new VietasrException(LastError("transcribe failed"));
            }
            return Result.FromJson(NativeMethods.PtrToString(raw));
        }

        public Session Stream(float sampleRate)
        {
            IntPtr sessionHandle = NativeMethods.vietasr_session_new(handle, sampleRate);
            if (sessionHandle == IntPtr.Zero)
            {
                throw new VietasrException(LastError("session creation failed"));
            }
            return new Session(sessionHandle);
        }

        public static IReadOnlyList<string> ListModules()
        {
            return ParseJsonStringArray(
                NativeMethods.PtrToString(NativeMethods.vietasr_list_modules()));
        }

        public static IReadOnlyList<string> ListPresets()
        {
            return ParseJsonStringArray(
                NativeMethods.PtrToString(NativeMethods.vietasr_list_presets()));
        }

        public static string Version()
        {
            return NativeMethods.PtrToString(NativeMethods.vietasr_version());
        }

        public static void SetLogLevel(int level)
        {
            NativeMethods.vietasr_set_log_level(level);
        }

        public void Dispose()
        {
            if (handle != IntPtr.Zero)
            {
                NativeMethods.vietasr_pipeline_free(handle);
                handle = IntPtr.Zero;
            }
        }

        private static string LastError(string fallback)
        {
            string e = NativeMethods.PtrToString(NativeMethods.vietasr_last_error());
            return string.IsNullOrEmpty(e) ? fallback : e;
        }

        private static IReadOnlyList<string> ParseJsonStringArray(string raw)
        {
            if (string.IsNullOrEmpty(raw))
            {
                return Array.Empty<string>();
            }
            try
            {
                using var doc = JsonDocument.Parse(raw);
                var list = new List<string>();
                foreach (var element in doc.RootElement.EnumerateArray())
                {
                    var s = element.GetString();
                    if (s != null)
                    {
                        list.Add(s);
                    }
                }
                return list;
            }
            catch (JsonException)
            {
                return Array.Empty<string>();
            }
        }
    }
}
